crates/task-impls/src/request.rs

```diff
@@ -160,237 +159,106 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>> TaskState for NetworkRequest
-            }
-        };
+            // start waiting
+            for _ in 0..4 {
+                let timeout = async_timeout(REQUEST_TIMEOUT, async {
+                    let mut response = None;
+                    while response.is_none() {
+                        let event = EventDependency::new(
+                            receiver.clone(),
+                            Box::new(move |event: &Arc<HotShotEvent<TYPES>>| {
+                                let event = event.as_ref();
+                                if let HotShotEvent::VidResponseRecv(proposal) = event {
+                                    proposal.data.view_number() == view
+                                } else {
+                                    false
+                                }
+                            }),
+                        )
+                        .completed()
+                        .await;
 
-        while !self.cancel_vid(&req).await {
-            match async_timeout(
-                REQUEST_TIMEOUT,
-                self.network
-                    .request_data::<TYPES>(serialized_msg.clone(), recipients_it.next().unwrap()),
-            )
-            .await
-            {
-                Ok(Ok(response)) => {
-                    match bincode::deserialize(&response) {
-                        Ok(ResponseMessage::Found(data)) => {
-                            self.handle_response_message(data).await;
-                            // keep trying, but expect the map to be populated, or view to increase
-                            async_sleep(REQUEST_TIMEOUT).await;
-                        }
-                        Ok(ResponseMessage::NotFound) => {
-                            info!("Peer Responded they did not have the data");
-                        }
-                        Ok(ResponseMessage::Denied) => {
-                            error!("Request for data was denied by the receiver");
-                        }
-                        Err(e) => {
-                            error!("Failed to deserialize response: {e}");
+                        if let Some(hs_event) = event.as_ref() {
+                            if let HotShotEvent::VidResponseRecv(proposal) = hs_event.as_ref() {
+                                response = Some(proposal.clone());
+                            }
                         }
                     }
-                }
-                Ok(Err(e)) => {
-                    warn!("Error Sending request.  Error: {:?}", e);
-                    async_sleep(REQUEST_TIMEOUT).await;
-                }
-                Err(_) => {
-                    warn!("Request to other node timed out");
-                }
-            }
-        }
-    }
-    /// Returns true if we got the data we wanted, or the view has moved on.
-    #[instrument(skip_all, target = "DelayedRequester", fields(id = self.id, view = *req.0))]
-    async fn cancel_vid(&self, req: &VidRequest<TYPES>) -> bool {
```

Comment:
I don't think we want to get rid of this, your changes basically means a task is spawned right away every time we get a proposal and never canceled until a node actually sends us the VID or it times out.  In the prior version we first delay, then check if we even need to send the request (and check again after each timeout) this means more often than not we don't request VID on the happy path.